samples: Bluetooth: Use bt_iso_chan_send_ts in iso_broadcast

cvinayak opened the pull request:
Use bt_iso_chan_send_ts API to send ISO data in the iso_broadcast sample.

The sample provides examples for the 4 statements in the Bluetooth Core Specification v6.0, Vol 6, Part G, Section 3.3 Time Stamp for SDU:

1. `should`: Time of arrival (use of `is_first_iso_data_send` and subsequent use of simulated `k_sleep()`)
```
When an HCI ISO Data packet sent by the Host does not contain a Time_Stamp or
the Time_Stamp value is not based on the Controller's clock, the Controller should
determine the CIS or BIS event to be used to transmit the SDU contained in that packet
based on the time of arrival of that packet.
``` 
2. `shall`: Time_Stamp corresponding to the Packet_Sequence_Number (verification of `ts` to match `tx_info.ts` for every `seq_num`)
```
In each direction, in a given packet, the Time_Stamp provided (if any) shall be one that
applies to the SDU corresponding to the Packet_Sequence_Number.
```
3. `should`: Compensate potential clock drift and jitter (calculation of `drift_comp_us`)
```
When using the Time_Stamp or another suitable method for synchronization, both the
higher layer and Controller should implement a mechanism to compensate potential
clock drift and jitter. These mechanisms should include long term drift compensation
and tolerances on the synchronization method used.
```
4. `may`: Use of timestamps at the Controller’s clock
```
The Controller or the higher layer sending or receiving an SDU may use any of the
following methods to transfer the synchronization reference of the SDU to the higher
layer:
• Timestamps at the Controller’s clock
• Any other signaling method suitable to the higher layer
```

It should also be noted that for a ISO Controller, use of time of arrival is a `should`, see Bluetooth Core Specification v6.0, Vol 6, Part B, Section 4.4.6.5 Broadcast Isochronous Data

```
If the source of the data fails to provide BN payloads for a BIS
event, bisPayloadCounter shall continue to increment as if it had provided the missing
payloads and the Link Layer should transmit empty PDUs.
```

Sample output:
```
d_00: @00:00:00.000000  *** Booting Zephyr OS build v4.0.0-3237-gc7d71fe8441b ***
d_00: @00:00:00.000000  Starting ISO Broadcast Demo
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.005768  [00:00:00.005,767] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 4.0 Build 99
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.005768  [00:00:00.005,767] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.006488  [00:00:00.006,469] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: Identity: ED:C7:E9:55:8A:4B (random)
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: HCI: version 5.4 (0x0d) revision 0x0000, manufacturer 0x05f1
d_00: @00:00:00.006488  [00:00:00.006,469] <inf> bt_hci_core: LMP: version 5.4 (0x0d) subver 0xffff
d_00: @00:00:00.007208  [00:00:00.007,202] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.007208  [00:00:00.007,202] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.007688  [00:00:00.007,659] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.008528  [00:00:00.008,514] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.008528  [00:00:00.008,514] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.009248  [00:00:00.009,246] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.009248  [00:00:00.009,246] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.010448  [00:00:00.010,437] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.010448  Waiting for BIG complete chan 0...
d_00: @00:00:00.016486  [00:00:00.016,479] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.016486  ISO Channel 0x80c9200 connected
d_00: @00:00:00.016486  [00:00:00.016,479] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.016486  ISO Channel 0x80c9214 connected
d_00: @00:00:00.016486  BIG create complete chan 0.
d_00: @00:00:00.016486  Waiting for BIG complete chan 1...
d_00: @00:00:00.016486  BIG create complete chan 1.
d_00: @00:00:00.016486  [00:00:00.016,479] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.016486  [00:00:00.016,479] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.016486  Sending value 0
d_00: @00:00:00.036530  [00:00:00.036,529] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.036530  Compensate 50 us drift
d_00: @00:00:00.036530  [00:00:00.036,529] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.036530  Sending value 1
d_00: @00:00:00.046509  [00:00:00.046,508] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.046509  [00:00:00.046,508] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.046509  Sending value 2
d_00: @00:00:00.056580  [00:00:00.056,579] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.056580  Compensate 128 us drift
d_00: @00:00:00.056580  [00:00:00.056,579] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.056580  Sending value 3
d_00: @00:00:00.066468  [00:00:00.066,467] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.066468  [00:00:00.066,467] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.066468  Sending value 4
d_00: @00:00:00.076539  [00:00:00.076,538] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.076539  Compensate 45 us drift
d_00: @00:00:00.076539  [00:00:00.076,538] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.076539  Sending value 5
d_00: @00:00:00.086518  [00:00:00.086,517] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.086518  [00:00:00.086,517] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.086518  Sending value 6
d_00: @00:00:00.096589  [00:00:00.096,588] <wrn> net_buf: Timeout discarded. No blocking in syswq
d_00: @00:00:00.096589  Compensate 145 us drift
```


Simulated trace of how Time-of-Arrival adjusted transmissions where the Pre-Transmissions are empty PDUs, but the actual ISO interval contains the SDU hence at a lower transport latency even when a Controller elects to use PTO:
[iso_broadcast_time_of_arrival.zip](https://github.com/user-attachments/files/18419375/iso_broadcast_time_of_arrival.zip)

cvinayak committed:
samples: Bluetooth: iso_broadcast: Use separate seq_num per channel

Use separate seq_num values for each channel so that it can
be synchronized with the Controller when streaming on a
channel starts delayed compared to other channels in the
same group.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

cvinayak committed:
samples: Bluetooth: Use bt_iso_chan_send_ts in iso_broadcast

Use bt_iso_chan_send_ts API to send ISO data in the
iso_broadcast sample.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -175,11 +179,22 @@ int main(void)
 		printk("BIG create complete chan %u.\n", chan);
 	}
 
+	uint32_t skip = 3U;
```
What is the purpose of this `skip` variable?

cvinayak replied:
This a separate commit to test that one stream starting later in time does synchronize to the same timestamp across all streams in the group.

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,14 +212,87 @@ int main(void)
+				/* Use sequence number 0, to synchronize current stream with the
+				 * current payload counter in the ISO group of streams.
+				 */
+				sn = 0U;
+
+			} else {
```
```suggestion
				sn = 0U;
			} else {
```

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,14 +212,87 @@ int main(void)
 			net_buf_add_mem(buf, iso_data, sizeof(iso_data));
-			ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, seq_num);
+
+			/* Ensure that first ISO data on all streams are given with sequence number
+			 * 0, so that the (Zephyr) Controller uses the same payload counter for
+			 * a given sequence number on all streams at an SDU interval.
+			 * Read Tx Sync interface returns error if no ISO data was given previously.
+			 */
+			ret = bt_iso_chan_get_tx_sync(&bis_iso_chan[chan], &tx_info);
+			if (ret) {
```
This may not work with all controllers. 

First reason:
The HCI command specifies 
> The Packet_Sequence_Number parameter contains the sequence number of a transmitted SDU.
At the time of calling `bt_iso_chan_get_tx_sync` here, we do not know if the SDU was transmitted. We may need to wait for the `sent` callback for that. 

Second reason:
Even in Core Spec 6.0 it still states 
> Host can use the HCI_LE_Read_ISO_TX_Sync command at any time to read the time stamp and packet sequence number of the last SDU **scheduled for transmission**.
in 4.7 LE Host and Controller synchronization. We have seen controllers return the PSN and timestamp of a transmitted SDU, and we have seen controllers return the PSN and timestamp of a scheduled transmission. 

I'm not sure how to best handle either of this complications

cvinayak replied:
I will go by Bluetooth Core Specification v6.0, Vol 4, Part E, Section 7.8.96 LE Read ISO TX Sync command

```
If the Host issues this command before an SDU has been transmitted by the Controller,
the Controller shall return the error code Command Disallowed (0x0C).
```

Regarding Bluetooth Core Specification v6.0, Vol 4, Part E, Section 4.7 LE Host and Controller synchronization

```
A Host can use the HCI_LE_Read_ISO_TX_Sync command at any time to read the
time stamp and packet sequence number of the last SDU scheduled for transmission.
```

I will go by `last SDU scheduled for transmission` because the Host which is not aware of the Controller clock will not know correctly when to call LE Read ISO TX Sync command such that `an SDU has been transmitted`.

As stated in the HCI Part of the specification, a conformant HCI Controller shall return `Command Disallowed (0x0C)` for LE Read ISO TX Sync command for the purpose used in this sample.

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,14 +212,87 @@ int main(void)
+			/* Send ISO data with timestamp, use sequence number 0 for first ISO data on
+			 * any stream.
+			 */
+			if (sn) {
+				ret = bt_iso_chan_send_ts(&bis_iso_chan[chan], buf, sn, ts);
+			} else {
+				ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, sn);
+			}
```
Checking for `sn == 0` won't work when it wraps.

cvinayak replied:
true, what do you suggest to use for checking if it is the first SDU being sent?

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -28,12 +28,16 @@ static K_SEM_DEFINE(sem_iso_data, CONFIG_BT_ISO_TX_BUF_COUNT,
 
 static uint16_t seq_num;
+static uint32_t ts;
+static int32_t  delta;
```
```suggestion
static int64_t  delta;
```

I think you need 64-bits for the delta, right? 
When the TS wraps from maximum to minimum then the delta will be nearly the maximum size of a uint32_t, which cannot be stored in a int32_t

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,14 +212,87 @@ int main(void)
+				return 0;
+			}
+
+			/* Timestamp delta with the Controller */
+			if (!delta) {
+				delta = tx_info.ts - ts;
+			} else {
+				int32_t drift = ((delta - tx_info.ts + ts) / BIG_SDU_INTERVAL_US) /
+						1000000U;
+
+				/* NOTE: As the `ts` is just a increase in software in this sample,
+				 *       we will not see any drift.
+				 */
+				if (drift) {
+					printk("Drift: %d ppm.\n", drift);
+				}
+
+				/* NOTE: Change in delta (drift) for a duration of SDU intervals can
+				 *       be used to synchronize the clock with the Controller.
+				 */
+			}
+
+			printk("Tx chan %u: seq num %u -> %u ts %u -> %u offset %u\n", chan,
+			       seq_num, tx_info.seq_num, ts, tx_info.ts, tx_info.offset);
+
+			/* Permit a reasonable clock jitter (at least Bluetooth active clock
+			 * jitter), in the verification of timestamp.
+			 */
+			uint32_t jitter_us = 2U;
+
+			/* Validate ISO data being enqueued at the required timestamp.
+			 * Ignore checking first ISO data sent.
+			 */
+			if (ts && sn &&
+			    !IN_RANGE((ts + delta), (tx_info.ts - jitter_us),
+				      (tx_info.ts + jitter_us))) {
+				printk("Mismatch in tx sync timestamp! (%u != %u)\n", ts,
+				       tx_info.ts);
+				return 0;
+			}
```
Suggest to move this part of a function instead. The loop has become quite complex with this addition

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,14 +212,87 @@ int main(void)
+				int32_t drift = ((delta - tx_info.ts + ts) / BIG_SDU_INTERVAL_US) /
+						1000000U;
+
+				/* NOTE: As the `ts` is just a increase in software in this sample,
+				 *       we will not see any drift.
+				 */
```
Not sure what you mean by that. 

In some previous on-target tests, I've seen the tx_info.ts change by a value that was not the SDU interval, but up to 100 microseconds in difference. Of course that is "just" jitter and not "drift".

cvinayak replied:
> Not sure what you mean by that.

This code down in this c file: `ts += BIG_SDU_INTERVAL_US;` i.e. ts is not sourced from a clock/timer but just an increase done in software.

Thalley reviewed (changes requested):
Keep in mind that this is a sample that should be kept as simple as possible, and that we don't run this sample in CI either. 

Consider applying similar changes to the actual ISO tests we run in BSIM

rugeGerritsen commented:
> Keep in mind that this is a sample that should be kept as simple as possible, and that we don't run this sample in CI either.
> 
> Consider applying similar changes to the actual ISO tests we run in BSIM

I agree with this point. If the purpose of this PR is to add validation of sending timestamped ISO data to the controller then we should rather create a test for it. Then we also do not have to care so much about controller specific behavior.

cvinayak commented:
> > Keep in mind that this is a sample that should be kept as simple as possible, and that we don't run this sample in CI either.
> > Consider applying similar changes to the actual ISO tests we run in BSIM
> 
> I agree with this point. If the purpose of this PR is to add validation of sending timestamped ISO data to the controller then we should rather create a test for it. Then we also do not have to care so much about controller specific behavior.

Currently, there is no sample using the `bt_iso_chan_send_ts`, `bt_bap_stream_send_ts`, `bt_cap_stream_send_ts` or `bt_iso_chan_get_tx_sync`. This PR addresses the need for a sample covering `bt_iso_chan_send_ts` API and the nitty-gritty around its use.

cvinayak commented:
@mringwal with reference to this discussion: https://interoperability.support.bluetooth.com/hc/en-us/community/posts/23958113843725-LE-Audio-via-HCI-and-ISO-Packet-Time-Stamps

Would you be able to comment/confirm if the calculated delta differences (example, in this sample) between the sent timestamp from Host and Read Tx Sync timestamp can be used to drift compensate the Host audio source clock (say, I2S sampling clock). I believe this is sufficient to supply ISO data at a constant rate in sync with the Controller's transmissions on-air.

(Thinking aloud, this is nothing new, this is similar to Bluetooth Peripheral device calculating the drift in its anchor point reception and accordingly drift compensates opening its receive window for subsequent connection intervals)

Thalley commented:
> Currently, there is no sample using the `bt_iso_chan_send_ts`, `bt_bap_stream_send_ts`, `bt_cap_stream_send_ts` or `bt_iso_chan_get_tx_sync`. This PR addresses the need for a sample covering `bt_iso_chan_send_ts` API and the nitty-gritty around its use.

I agree that it's useful to have a sample using `bt_iso_chan_send_ts`, but my worry is that _how_ the application originally obtains an initial timestamp to "start from" is controller dependent. So the approach used here might work with ZLL, and might not work with SDC, PCFT or an Intel controller on your PC, given the 2 different definitions of what the command returns. This is also not taking into account that HCI does not have any timing requirements. 

We can use the approach here, but perhaps without the validation, and some comments that this may not work for all controllers and that it is best-effort.

cvinayak commented:
> my worry is that how the application originally obtains an initial timestamp to "start from" is controller dependent.

Implementation of timestamp in the application (`ts`) is independent of the Controller (`tx_info.ts`), application uses a free running timestamp and Controller has its own free running reference clock timestamp. There is nothing Controller specific in this sample, and adheres to the Bluetooth Specifications definition of LE Read ISO TX Sync command.

mringwal commented:
> @mringwal with reference to this discussion: https://interoperability.support.bluetooth.com/hc/en-us/community/posts/23958113843725-LE-Audio-via-HCI-and-ISO-Packet-Time-Stamps
> 
> Would you be able to comment/confirm if the calculated delta differences (example, in this sample) between the sent timestamp from Host and Read Tx Sync timestamp can be used to drift compensate the Host audio source clock (say, I2S sampling clock).

Thanks for asking. If we can get the both the Controller timestamp as well as SDU Sync Reference resp. the  Read Tx Sync with a an error of a few microseconds, it should be possible for the Host to properly do local playback of sent audio (in case of audio forwarding) or received (e.g. True Wireless earbuds). I've use "possible" because we don't have a working implementation for this yet, we're using an STM32F4 dev kit with an RF5340 DK as platform at the moment. 

However, during GMAP testing, I can confirm (i.e. measured with a logic analyzer) that I got the SDU Sync Ref on two nRF5340 with an error of +/- 2 us and similar values for the Read TX Sync + Transport Latency of the sender vs. SDU Sync Latency on the unicast receiver. All was done with SoftDevice

cvinayak committed:
samples: Bluetooth: iso_broadcast: Drift compensation of source

Example of drift compensation of simulated ISO data source
with respect kernel system clock drift.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

rugeGerritsen commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -175,11 +175,29 @@ int main(void)
 		printk("BIG create complete chan %u.\n", chan);
 	}
 
+	/* Simulate delayed streaming on one stream */
+	const uint8_t skip_chan = UINT8_MAX;
+	uint32_t skip = 5U;
```
I would be happy if we could remove testing code out from the sample. If we want to document how to do these kinds of things we could consider documenting it in the readme.

cvinayak replied:
Sure will remove this commit that adds the check in skipped ISO data send.

rugeGerritsen commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,22 +215,116 @@ int main(void)
 		iso_send_count++;
-		seq_num++;
+
+		uint32_t sleep_us = BIG_SDU_INTERVAL_US - (drift_comp_us * 2);
+
+		drift_comp_us = 0U;
+
+		uint32_t ts_prev = k_ticks_to_us_floor32(sys_clock_tick_get());
+
+		k_sleep(K_USEC(sleep_us));
+
+		uint32_t ts_curr = k_ticks_to_us_floor32(sys_clock_tick_get());
+
+		ts += MAX(sleep_us, (ts_curr - ts_prev));
```
Why are we adding sleep code the sample? When using timestamps there is no need to sleep, the controller would know when to send the payloads based upon the received timestamp

cvinayak replied:
The code here is simulating the relative drift in the sampling clock with respect to the Controller reference clock. And expects the Controller to handle the below specification statement:

```
When an HCI ISO Data packet sent by the Host does not contain a Time_Stamp or
the Time_Stamp value is not based on the Controller's clock, the Controller should
determine the CIS or BIS event to be used to transmit the SDU contained in that packet
based on the time of arrival of that packet.
```

The `k_sleep` is used to simulate the time of arrival of ISO data packets to the Controller.

rugeGerritsen replied:
But if we are using timestamps, then time of arrival is likely not used anyways. Personally I would therefore remove this to reduce the complexity of the sample

rugeGerritsen commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,22 +215,116 @@ int main(void)
 			net_buf_add_mem(buf, iso_data, sizeof(iso_data));
-			ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, seq_num);
+
+			/* Read Tx Sync interface returns error if no ISO data was given previously.
+			 */
```
Can we avoid calling `bt_iso_chan_get_tx_sync()` when no data has been sent on a channel? Then we avoid error prints in the sample

rugeGerritsen commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,22 +215,116 @@ int main(void)
+			printk("Tx chan %u: seq num %u -> %u ts %u -> %u offset %u\n", chan,
+			       seq_num[chan], tx_info.seq_num, ts, tx_info.ts, tx_info.offset);
+
+			/* Permit a reasonable clock jitter (at least Bluetooth active clock
+			 * jitter), in the verification of timestamp.
+			 */
+			uint32_t jitter_us = BIG_SDU_INTERVAL_US;
```
Allowing a jitter of an entire SDU interval seems a bit too much :)

cvinayak replied:
The `k_sleep()` (based on kernel system clock) is very flakky, and has large jitters!

rugeGerritsen commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,22 +215,116 @@ int main(void)
+			if (ret) {
+				printk("Failed to get tx sync before send (%d)\n", ret);
+
+				/* Use sequence number 0, to synchronize current stream with the
+				 * current payload counter in the ISO group of streams.
+				 */
```
The spec doesn't state that the sequence number shall be used to find the corresponding payload counter. The controller is free to ignore it completely or only ignore it when it seems wrong.

I would therefore consider updating the comment:
```suggestion
				/* Use sequence number 0 for the first SDU sent on a given stream.
				 */
```
More information could be put in the readme of the sample.

rugeGerritsen reviewed (changes requested):
I would also be happy if you could update the commit messages to explain why we are adding the changes in each commit. See commit guidelines for more details

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,23 +206,117 @@ int main(void)
+
+			/* Timestamp delta with the Controller */
+			if (!delta) {
+				delta = tx_info.ts - ts;
```
Do we need a cast here? If `ts > tx_info.ts`, does it correctly become a negative value, or does it just underflow before the assignment?

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,23 +206,117 @@ int main(void)
 				return 0;
 			}
 
+			/* Get Tx Sync, to verify that all streams have their ISO data at the same
```
Still suggest to move all code related to TX sync to a new function, to reduce the length of this loop.

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,23 +206,117 @@ int main(void)
+
+			/* Use one channel timestamp to accumulate relative drift */
+			} else if (chan == 0U) {
+				/* NOTE: Change in delta (drift) for a duration of SDU intervals can
+				 *       be used to synchronize the clock with the Controller.
+				 */
+				static uint8_t countdown = SOURCE_DRIFT_COMP_COUNTDOWN;
+				static int64_t drift_us;
+
+				/* Accumulate relative drift */
+				drift_us += delta - tx_info.ts + ts;
+
+				/* Transfer the accumulated drift to compensate sampling interval */
+				if (!countdown--) {
+					countdown = SOURCE_DRIFT_COMP_COUNTDOWN;
+					drift_comp_us = drift_us;
+					drift_us = 0U;
+
+					printk("Compensate %lld us drift\n", drift_comp_us);
+				}
+			}
+
+			/* Permit a reasonable clock jitter (at least Bluetooth active clock
+			 * jitter), in the verification of timestamp.
+			 *
+			 * Lets consider that the source clock is allowed to drift up to the
+			 * BIG SDU interval before drift compensation is applied.
+			 */
+			uint32_t jitter_us = BIG_SDU_INTERVAL_US;
+
+			/* Validate ISO data being enqueued at the required timestamp.
+			 * Ignore checking first ISO data sent.
+			 */
+			if (!is_first_iso_data_send &&
+			    !IN_RANGE(tx_info.ts, (ts + delta - jitter_us),
+				      (ts + delta + jitter_us))) {
+				printk("Mismatch in tx sync timestamp! (%llu != %u)\n",
+				       (ts + delta), tx_info.ts);
+				return 0;
+			}
```
I don't understand the purpose of this. 

Why does the application care about the drift? 

If I have an application and I know the timestamp of the SDU, I should just be able to increment hte timestamp with the SDU interval. Jitter and drift should not come into play here.

cvinayak replied:
Refer to https://github.com/zephyrproject-rtos/zephyr/pull/83139#discussion_r1894764005, about time of arrival, drift and jitter stated in the Bluetooth Core Specification.

Thalley commented on samples/bluetooth/iso_broadcast/src/main.c:
```diff
@@ -197,23 +206,117 @@ int main(void)
-		seq_num++;
 
+
+		/* NOTE: Here we are simulating a drift compensated sleep for
+		 *       BIG_SDU_INTERVAL_US
+		 */
+		uint32_t sleep_us;
+
+		if (drift_comp_us < BIG_SDU_INTERVAL_US) {
+			sleep_us = BIG_SDU_INTERVAL_US - drift_comp_us;
+		} else {
+			sleep_us = 0U;
+		}
+
+		/* Reset drift compensation value */
+		drift_comp_us = 0;
+
+		/* Sleep, simulating source ISO data sampling */
+		uint32_t ts_curr;
+
+		do {
+			k_sleep(K_USEC(10));
+			ts_curr = k_ticks_to_us_floor32(sys_clock_tick_get());
+		} while ((ts_curr - ts_prev) < sleep_us);
+
+		/* Increment the application timestamp by the simulated sampling duration */
+		ts += (ts_curr - ts_prev);
+		ts_prev = ts_curr;
```
Ditto here: What is the purpose here? 

What's the reason to have a sleep here?

cvinayak replied:
The sample provides examples for the 4 statements in the Bluetooth Core Specification v6.0, Vol 6, Part G, Section 3.3 Time Stamp for SDU:

1. `should`: Time of arrival (use of `is_first_iso_data_send` and subsequent use of simulated `k_sleep()`)
```
When an HCI ISO Data packet sent by the Host does not contain a Time_Stamp or
the Time_Stamp value is not based on the Controller's clock, the Controller should
determine the CIS or BIS event to be used to transmit the SDU contained in that packet
based on the time of arrival of that packet.
``` 
2. `shall`: Time_Stamp corresponding to the Packet_Sequence_Number (verification of `ts` to match `tx_info.ts` for every `seq_num`)
```
In each direction, in a given packet, the Time_Stamp provided (if any) shall be one that
applies to the SDU corresponding to the Packet_Sequence_Number.
```
3. `should`: Compensate potential clock drift and jitter (calculation of `drift_comp_us`)
```
When using the Time_Stamp or another suitable method for synchronization, both the
higher layer and Controller should implement a mechanism to compensate potential
clock drift and jitter. These mechanisms should include long term drift compensation
and tolerances on the synchronization method used.
```
4. `may`: Use of timestamps at the Controller’s clock
```
The Controller or the higher layer sending or receiving an SDU may use any of the
following methods to transfer the synchronization reference of the SDU to the higher
layer:
• Timestamps at the Controller’s clock
• Any other signaling method suitable to the higher layer
```

Thalley replied:
The spec states `When an HCI ISO Data packet sent by the Host does not contain a Time_Stamp or the Time_Stamp value is not based on the Controller's clock`, but since you are using the timestamp from the TX sync, you are using a timestamp based on the controller's clock, right? 

Since it's based on the controller's clock, and that you can just increment with the SDU interval, I still don't understand why you need to care about drift compensation or jitter. 

The timestamp should always be incremented by the SDU interval, so everything you are doing here seems to only be useful for the sleep, which makes some sense but not quite IMO. Having an accurate sleep in the application layer won't really ensure anything, as there is a indeterminable time from the application calling `bt_iso_chan_send_ts` to the HCI command being received by the controller. 

I might be missing something obvious here, and it might be easier to understand once the code has been moved to helper functions to better understand each step.

cvinayak replied:
> I still don't understand why you need to care about drift compensation or jitter.

It is the time of arrival, audio sampling clock vs Bluetooth clock. ISO interval of 10 ms at Bluetooth clock versus 10 ms interval frame given to the Controller have to be in sync, else there will be under-run or over-run.

How Analog Devices mentioned once: "marching to the beat of the same drummer" ! https://www.analog.com/media/en/analog-dialogue/volume-28/number-1/articles/volume28-number1.pdf#page=9

Thalley replied:
> It is the time of arrival, audio sampling clock vs Bluetooth clock.

Doesn't really make sense to talk about audio sampling in a generic ISO sample IMO. That something that makes more sense to do in the audio samples. 

By ISO interval here, do you mean the SDU interval? For LE Audio, the host sends 1 to many audio frames per SDU (and thus per SDU interval, and not ISO interval). 

If we defer to whether buffers are available, then we have the following 2 cases:
1) Bluetooth (host/app) receives more audio frames than we get buffers from the controller to send. Only option is to drop audio frames when that happen (with some buffering possible of course)
2) Bluetooth (host/app) receive fewer audio frames than we get buffers from the controller. In this case the host just won't send anything, because what can we send? 

What the application does in either of those cases is very dependent on the system, and IMO difficult to reasonably implement in a sample application, and even more so when we are at the ISO level, where there are even less context about the type of data. I'm no audio expert, so please excuse me if I'm overseeing something obvious :)

Thalley reviewed (commented):
I still feel that the majority of the changes in this PR are not relevant for a sample, and looks more like testing code, that seem more relevant to a babblesim test, or alternatively added to the benchmark samples (that are closer to tests as they are benchmarks, and not simple samples).

cvinayak commented:
> I still feel that the majority of the changes in this PR are not relevant for a sample, and looks more like testing code, that seem more relevant to a babblesim test, or alternatively added to the benchmark samples (that are closer to tests as they are benchmarks, and not simple samples).

This is not a test, but a sample use of `bt_iso_chan_send_ts` to meet the 4 statements in Bluetooth Core Specification v6.0, Vol 6, Part G, Section 3.3 Time Stamp for SDU, as commented here https://github.com/zephyrproject-rtos/zephyr/pull/83139#discussion_r1894764005

cvinayak commented:
> > @mringwal with reference to this discussion: https://interoperability.support.bluetooth.com/hc/en-us/community/posts/23958113843725-LE-Audio-via-HCI-and-ISO-Packet-Time-Stamps
> > Would you be able to comment/confirm if the calculated delta differences (example, in this sample) between the sent timestamp from Host and Read Tx Sync timestamp can be used to drift compensate the Host audio source clock (say, I2S sampling clock).
> 
> Thanks for asking. If we can get the both the Controller timestamp as well as SDU Sync Reference resp. the Read Tx Sync with a an error of a few microseconds, it should be possible for the Host to properly do local playback of sent audio (in case of audio forwarding) or received (e.g. True Wireless earbuds). I've use "possible" because we don't have a working implementation for this yet, we're using an STM32F4 dev kit with an RF5340 DK as platform at the moment.
> 
> However, during GMAP testing, I can confirm (i.e. measured with a logic analyzer) that I got the SDU Sync Ref on two nRF5340 with an error of +/- 2 us and similar values for the Read TX Sync + Transport Latency of the sender vs. SDU Sync Latency on the unicast receiver. All was done with SoftDevice

@mringwal  This sample illustrates an ISO source, where two clocks are in use, ISO data sampling clock and the Controller free running reference clock.  In this case the HCI LE Read ISO Tx Sync command is used for drift compensation of Time of Arrival for the transmission of ISO data.

For an ISO sink (unframed ISO data), there are three clocks in theory, peer ISO sampling clock, local Controller free running clock, and ISO sink playback clock. There is drift in consecutive received ISO data Time_Stamp (drift between local and remote Controller), and a corresponding drift in Time of Arrival for playback (drift between playback clock and peer ISO sampling clock).
The local playback sink clock will need to be drift compensated (or the received ISO data run through a re-sampler when mixing ISO data from multiple peer source).

Thinking aloud, `iso_receive` sample can be updated with a sink thread consuming the received ISO data using the kernel system clock as the playback clock; the thread can simulate a drift compensated SDU interval sleep duration, and validate for buffer under-run and over-run during the ISO data reception.

cvinayak committed:
samples: Bluetooth: iso_broadcast: Adjusting of Time-of-Arrival

Adjusting of Time-of-Arrival to demonstrate reduced
transport latency when a Controller has elected to use PTO
for cases where ISO interval is SDU interval.

Signed-off-by: Vinayak Kariappa Chettimada <vich@nordicsemi.no>

mringwal commented:
I would like the stale label removed. The example tries to document not-so-well-documented behavior/expectations on the Controller in the Bluetooth Specification.
